node/src/components/linear_chain_sync/operations.rs

```diff
@@ -113,6 +113,66 @@ where
     Ok(block_header)
 }
 
+/// Get trusted block headers; falls back to genesis if none are available
+async fn get_trusted_validator_weights<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    chainspec: &Chainspec,
+    trusted_header: &BlockHeader,
+) -> Result<BTreeMap<PublicKey, U512>, LinearChainSyncError<I>>
+where
+    REv: From<FetcherRequest<I, BlockHeader>> + From<NetworkInfoRequest<I>> + From<StorageRequest>,
+    I: Eq + Debug + Clone + Send + 'static,
+{
+    // If we are right after genesis, use the genesis validators.
+    if trusted_header.era_id() == EraId::new(0) {
+        let genesis_validator_weights: BTreeMap<PublicKey, U512> = chainspec
+            .network_config
+            .accounts_config
+            .accounts()
+            .iter()
+            .filter_map(|account_config| {
+                if account_config.is_genesis_validator() {
+                    Some((
+                        account_config.public_key(),
+                        account_config.bonded_amount().value(),
+                    ))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        return Ok(genesis_validator_weights);
+    }
+
+    // Check that we are not restarting right after an emergency restart, which is too early
+    // Consider emitting a switch block after an emergency restart to make this simpler...
+    let maybe_last_emergency_restart_era_id = chainspec.protocol_config.last_emergency_restart;
+
+    let min_era = maybe_last_emergency_restart_era_id.unwrap_or_else(|| EraId::new(0));
+    if min_era >= trusted_header.era_id() {
+        return Err(LinearChainSyncError::TrustedHeaderEraTooEarly {
+            trusted_header: Box::new(trusted_header.clone()),
+            maybe_last_emergency_restart_era_id,
+        });
+    };
+
+    // Fetch each parent hash one by one until we have the trusted validator weights
+    let mut current_header_to_walk_back_from = Box::new(trusted_header.clone());
+    loop {
+        if let Some(trusted_validator_weights) =
+            current_header_to_walk_back_from.next_era_validator_weights()
+        {
+            return Ok(trusted_validator_weights.clone());
+        }
+        current_header_to_walk_back_from = fetch_and_store_block_header(
+            effect_builder,
+            *current_header_to_walk_back_from.parent_hash(),
+        )
+        .await?;
+    }
+}
```

Comment:
Doesn't this duplicate [`Effect::get_era_validators`](https://github.com/casper-network/casper-node/blob/5dda5052bd1265de9164d7d3be45d24750686697/node/src/effect.rs#L1557), except that that retrieves the validators from the global state? But at that point, we should have the global state available so it should work, shouldn't it?

Reply:
Almost, but not exactly.  `Effect::get_era_validators` can does storage and contract-runtime requests.  We don't have anything in either storage or the state trie at the point this code is called.  The functions are similar, however: they both fallback to the chainspec in the special case where we are asking for the validators for era 0.